content/post/github_topic_modeling.md

```diff
@@ -0,0 +1,273 @@
+Each "document" is represented by a bag-of-words model, that is, a set of
+occurring words and their frequencies. Since I am Russian, I had the introduction
+to topic modeling through the awesome lectures by [Dr. Vorontsov](http://www.machinelearning.ru/wiki/index.php?title=%D0%A3%D1%87%D0%B0%D1%81%D1%82%D0%BD%D0%B8%D0%BA:Vokov)
+at Yandex's School for Data Analysis [PDF](http://www.machinelearning.ru/wiki/images/e/e6/Voron-ML-TopicModeling-slides.pdf).
+There exist different models to do topic modeling, the most famous (but not the best) being
+[Latent Dirichlet Allocation](https://en.wikipedia.org/wiki/Latent_Dirichlet_allocation) (LDA).
+Dr. Vorontsov managed to generalize all possible bag-of-words-based topic models
+into [Additive Regularization of Topic Models methodology](http://link.springer.com/article/10.1007/s10994-014-5476-6) (ARTM).
+Thus LDA becomes a subset of possible ARTMs. The really cool thing about ARTM is
+that Dr. Vorontsov's PHd students developed the proof-of-concept software and it
+is open source: [bigartm/bigartm](https://github.com/bigartm/bigartm).
+
+Normally, topic modeling is applied to ordinary text documents. We at source{d}
+struggle with huge amount of data originated from GitHub repositories (an soon
+  we'll handle every open source repository in the world). A natural idea is
+to treat every repository as a bag-of-words and apply the powerful engine of
+topic modeling of Dr. Vorontsov. For example, we could extract names from the
+source code and consider them "words". This article elaborates in this direction,
+presents the results and explains how to repeat them.
+
+![topic cloud](wordcloud.png)
+
+TL;DR:
+```
+docker run srcd/github_topics apache/spark
+```
+(replace `apache/spark` with any GitHub repository you wish to analyze).
+
+[OpenDocument spreadsheet with the extracted topics](https://storage.googleapis.com/github-repositories-topic-modeling/topics.ods).
+
+[JSON with the extracted topics](https://storage.googleapis.com/github-repositories-topic-modeling/topics.json).
+
+[Trained model](https://storage.googleapis.com/github-repositories-topic-modeling/repo_topic_modeling.pickle.gz) - 40M, gzipped pickle for Python 3.4+, Pandas 1.18+.
+
+Theory
+------
+Suppose that we have a topic probabilistic model of the collection of documents
+\\(D\\) which describes the occurrence of terms \\(w\\) in documents \\(d\\) with topics \\(t\\):
+$$
+p(w|d) = \\sum_{t\\in T} p(w|t) p(t|d)
+$$
+Here \\(p(w|t)\\) is the probability of the term \\(w\\) to belong to the topic \\(t\\),
+\\(p(t|d)\\) is the probability of the topic \\(t\\) to belong to the document \\(d\\),
+thus the whole formula is just the expression of [total probability](https://en.wikipedia.org/wiki/Law_of_total_probability).
+We'd like to solve the problem of recovering \\(p(w|t)\\) and \\(p(t|d)\\) from
+the given set of documents \\(\\left\\{d\\in D: d = \\left\\{w\_ 1 \\dots w\_ {n\_ d}\\right\\}, |d| = n\_ d\\right\\}\\). The terms
```

Comment:
Remove the definition of $|d|$, we are not using it for anything and it is generally understood that $n_d$ will be the cardinality of $d$.